File: scripts/proto-smoke.py
#!/usr/bin/env python3
"""M3a feasibility smoke: drive the patched hax over inherited fds (no scraping).

Phase 1 (lifecycle): send a `submit` and assert the EXACT ordered M3a sequence
  ready, user_turn_started, assistant_turn_started, assistant_delta(>=1),
  assistant_turn_finished{content}, idle — with deltas concatenating to content.
Phase 2 (interrupt): with a slow mock script, submit, then `interrupt` the live
  turn and assert it aborts back to idle promptly without delivering the full
  scripted text.

Exits non-zero on any mismatch.
"""
import json
import os
import subprocess
import sys
import tempfile
import time

HAX = os.path.join(os.path.dirname(__file__), "..", "vendor", "hax", "build", "hax")


class Engine:
    """Spawn hax with protocol fds wired and read/write JSONL over them."""

    def __init__(self, extra_env=None):
        self.ev_r, ev_w = os.pipe()
        ctl_r, self.ctl_w = os.pipe()
        env = {**os.environ, "HAX_PROVIDER": "mock", "HAX_NO_SESSION": "1"}
        if extra_env:
            env.update(extra_env)
        self.proc = subprocess.Popen(
            [HAX, f"--protocol-fd={ev_w}", f"--control-fd={ctl_r}"],
            pass_fds=(ev_w, ctl_r),
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            env=env,
        )
        os.close(ev_w)
        os.close(ctl_r)
        self._buf = b""

    def send(self, control):
        os.write(self.ctl_w, json.dumps(control).encode() + b"\n")

    def read_until(self, types, timeout=10.0):
        """Read events until one whose type is in `types`; return all collected."""
        import select

        events = []
        deadline = time.monotonic() + timeout
        while True:
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                raise TimeoutError(f"no {types} within {timeout}s; got {events}")
            r, _, _ = select.select([self.ev_r], [], [], remaining)
            if not r:
                continue
            chunk = os.read(self.ev_r, 4096)
            if not chunk:
                return events
            self._buf += chunk
            while b"\n" in self._buf:
                line, self._buf = self._buf.split(b"\n", 1)
                if not line.strip():
                    continue
                ev = json.loads(line)
                events.append(ev)
                if ev.get("type") in types:
                    return events

    def close(self):
        try:
            os.close(self.ctl_w)
        except OSError:
            pass
        try:
            self.proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            self.proc.kill()
        os.close(self.ev_r)


def phase_lifecycle(failures):
    eng = Engine()
    try:
        ready = eng.read_until({"ready"})[-1]
        if ready.get("type") != "ready" or "protocol" not in ready:
            failures.append(f"bad ready: {ready}")
        eng.send({"type": "submit", "text": "say hello"})
        evs = eng.read_until({"idle"})
        seq = [e["type"] for e in evs]
        print(f"  lifecycle: ready + {seq}")

        # Exact required ordered subsequence (deltas required).
        required = [
            "user_turn_started",
            "assistant_turn_started",
            "assistant_delta",
            "assistant_turn_finished",
            "idle",
        ]
        pos = 0
        for t in seq:
            if pos < len(required) and t == required[pos]:
                pos += 1
        if pos != len(required):
            failures.append(f"sequence missing/oo-order; need {required} as subseq of {seq}")

        deltas = [e for e in evs if e["type"] == "assistant_delta"]
        if not deltas:
            failures.append("no assistant_delta emitted (required in M3a)")
        finished = next((e for e in evs if e["type"] == "assistant_turn_finished"), None)
        content = finished.get("content", "") if finished else ""
        if "say hello" not in content:
            failures.append(f"content did not reflect input: {content!r}")
        joined = "".join(e.get("text", "") for e in deltas)
        if joined.strip() != content.strip():
            failures.append(f"delta concat {joined!r} != content {content!r}")
    finally:
        eng.close()


def phase_interrupt(failures):
    script = tempfile.NamedTemporaryFile("w", suffix=".mock", delete=False)
    # One slow turn: a 2.5s delay before any text. The agent's stream tick polls
    # the control fd every ~50ms during the delay, so `interrupt` aborts it.
    script.write("delay 2500\ntext THIS_SHOULD_NOT_APPEAR\nend-turn\n")
    script.close()
    eng = Engine(extra_env={"HAX_MOCK_SCRIPT": script.name})
    try:
        eng.read_until({"ready"})
        eng.send({"type": "submit", "text": "go"})
        # Wait until the assistant turn is live, then interrupt it.
        eng.read_until({"assistant_turn_started"})
        t0 = time.monotonic()
        eng.send({"type": "interrupt"})
        evs = eng.read_until({"idle"}, timeout=5.0)
        elapsed = time.monotonic() - t0
        print(f"  interrupt: returned to idle in {elapsed:.2f}s; {[e['type'] for e in evs]}")
        if elapsed >= 2.0:
            failures.append(f"interrupt did not abort promptly ({elapsed:.2f}s ~ full 2.5s delay)")
        if any("THIS_SHOULD_NOT_APPEAR" in e.get("text", "") for e in evs):
            failures.append("interrupted turn still delivered the scripted text")
        if not any(e["type"] == "idle" for e in evs):
            failures.append("no idle after interrupt")
    finally:
        eng.close()
        os.unlink(script.name)


def main():
    if not os.path.exists(HAX):
        print(f"FAIL: hax binary not built at {HAX}", file=sys.stderr)
        return 1
    failures = []
    phase_lifecycle(failures)
    phase_interrupt(failures)
    if failures:
        print("PROTO SMOKE FAIL:")
        for f in failures:
            print(f"  - {f}")
        return 1
    print("PROTO SMOKE PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/proto-smoke.mjs
#!/usr/bin/env node
/**
 * M3a feasibility smoke: drive the patched hax over inherited fds (no scraping).
 *
 * Phase 1 (lifecycle): send a `submit` and assert the EXACT ordered M3a sequence
 *   ready, user_turn_started, assistant_turn_started, assistant_delta(>=1),
 *   assistant_turn_finished{content}, idle — with deltas concatenating to content.
 * Phase 2 (interrupt): with a slow mock script, submit, then `interrupt` the live
 *   turn and assert it aborts back to idle promptly without delivering the full
 *   scripted text.
 *
 * Exits non-zero on any mismatch. (Port of the original proto-smoke.py; the
 * child fds are deterministic here — stdio positions 3/4 — so no pass_fds
 * bookkeeping is needed.)
 */
import { spawn } from "node:child_process";
import { existsSync, mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const HAX = join(dirname(fileURLToPath(import.meta.url)), "..", "vendor", "hax", "build", "hax");

/** Spawn hax with protocol fds wired and read/write JSONL over them. */
class Engine {
	constructor(extraEnv = {}) {
		this.proc = spawn(HAX, ["--protocol-fd=3", "--control-fd=4"], {
			stdio: ["ignore", "ignore", "ignore", "pipe", "pipe"],
			env: { ...process.env, HAX_PROVIDER: "mock", HAX_NO_SESSION: "1", ...extraEnv },
		});
		this.events = this.proc.stdio[3];
		this.controls = this.proc.stdio[4];
		this.queue = [];
		this.waiters = [];
		this.eof = false;
		let buf = "";
		this.events.setEncoding("utf8");
		this.events.on("data", (chunk) => {
			buf += chunk;
			let nl;
			while ((nl = buf.indexOf("\n")) >= 0) {
				const line = buf.slice(0, nl).trim();
				buf = buf.slice(nl + 1);
				if (!line) continue;
				const ev = JSON.parse(line);
				const w = this.waiters.shift();
				if (w) w(ev);
				else this.queue.push(ev);
			}
		});
		this.events.on("end", () => {
			this.eof = true;
			for (const w of this.waiters.splice(0)) w(null);
		});
	}

	send(control) {
		this.controls.write(`${JSON.stringify(control)}\n`);
	}

	/** Read events until one whose type is in `types`; return all collected. */
	async readUntil(types, timeoutMs = 10_000) {
		const events = [];
		const deadline = Date.now() + timeoutMs;
		for (;;) {
			const ev = await this.next(deadline - Date.now(), types, events);
			if (ev === null) return events; // EOF
			events.push(ev);
			if (types.has(ev.type)) return events;
		}
	}

	next(ms, types, events) {
		if (this.queue.length) return Promise.resolve(this.queue.shift());
		if (this.eof) return Promise.resolve(null);
		return new Promise((resolve, reject) => {
			const waiter = (ev) => {
				clearTimeout(timer);
				resolve(ev);
			};
			const timer = setTimeout(() => {
				const i = this.waiters.indexOf(waiter);
				if (i >= 0) this.waiters.splice(i, 1); // no orphan steals later events
				const got = events.map((e) => e.type);
				reject(new Error(`no ${[...types]} within timeout; got ${JSON.stringify(got)}`));
			}, Math.max(1, ms));
			this.waiters.push(waiter);
		});
	}

	async close() {
		this.controls.end();
		const exited = new Promise((resolve) => {
			if (this.proc.exitCode !== null) return resolve();
			this.proc.once("exit", resolve);
		});
		const timer = setTimeout(() => this.proc.kill("SIGKILL"), 5000);
		await exited;
		clearTimeout(timer);
	}
}

async function phaseLifecycle(failures) {
	const eng = new Engine();
	try {
		const ready = (await eng.readUntil(new Set(["ready"]))).at(-1);
		if (ready?.type !== "ready" || !("protocol" in ready)) {
			failures.push(`bad ready: ${JSON.stringify(ready)}`);
		}
		eng.send({ type: "submit", text: "say hello" });
		const evs = await eng.readUntil(new Set(["idle"]));
		const seq = evs.map((e) => e.type);
		console.log(`  lifecycle: ready + ${JSON.stringify(seq)}`);

		// Exact required ordered subsequence (deltas required).
		const required = [
			"user_turn_started",
			"assistant_turn_started",
			"assistant_delta",
			"assistant_turn_finished",
			"idle",
		];
		let pos = 0;
		for (const t of seq) if (pos < required.length && t === required[pos]) pos += 1;
		if (pos !== required.length) {
			failures.push(`sequence missing/oo-order; need ${required} as subseq of ${seq}`);
		}

		const deltas = evs.filter((e) => e.type === "assistant_delta");
		if (deltas.length === 0) failures.push("no assistant_delta emitted (required in M3a)");
		const finished = evs.find((e) => e.type === "assistant_turn_finished");
		const content = finished?.content ?? "";
		if (!content.includes("say hello")) {
			failures.push(`content did not reflect input: ${JSON.stringify(content)}`);
		}
		const joined = deltas.map((e) => e.text ?? "").join("");
		if (joined.trim() !== content.trim()) {
			failures.push(
				`delta concat ${JSON.stringify(joined)} != content ${JSON.stringify(content)}`,
			);
		}
	} finally {
		await eng.close();
	}
}

async function phaseInterrupt(failures) {
	const dir = mkdtempSync(join(tmpdir(), "proto-smoke-"));
	const script = join(dir, "slow.mock");
	// One slow turn: a 2.5s delay before any text. The agent's stream tick polls
	// the control fd every ~50ms during the delay, so `interrupt` aborts it.
	writeFileSync(script, "delay 2500\ntext THIS_SHOULD_NOT_APPEAR\nend-turn\n");
	const eng = new Engine({ HAX_MOCK_SCRIPT: script });
	try {
		await eng.readUntil(new Set(["ready"]));
		eng.send({ type: "submit", text: "go" });
		// Wait until the assistant turn is live, then interrupt it.
		await eng.readUntil(new Set(["assistant_turn_started"]));
		const t0 = Date.now();
		eng.send({ type: "interrupt" });
		const evs = await eng.readUntil(new Set(["idle"]), 5000);
		const elapsed = (Date.now() - t0) / 1000;
		const seq = evs.map((e) => e.type);
		console.log(`  interrupt: returned to idle in ${elapsed.toFixed(2)}s; ${JSON.stringify(seq)}`);
		if (elapsed >= 2.0) {
			failures.push(`interrupt did not abort promptly (${elapsed.toFixed(2)}s ~ full 2.5s delay)`);
		}
		if (evs.some((e) => (e.text ?? "").includes("THIS_SHOULD_NOT_APPEAR"))) {
			failures.push("interrupted turn still delivered the scripted text");
		}
		if (!evs.some((e) => e.type === "idle")) failures.push("no idle after interrupt");
	} finally {
		await eng.close();
		rmSync(dir, { recursive: true, force: true });
	}
}

async function main() {
	if (!existsSync(HAX)) {
		console.error(`FAIL: hax binary not built at ${HAX}`);
		return 1;
	}
	const failures = [];
	await phaseLifecycle(failures);
	await phaseInterrupt(failures);
	if (failures.length) {
		console.log("PROTO SMOKE FAIL:");
		for (const f of failures) console.log(`  - ${f}`);
		return 1;
	}
	console.log("PROTO SMOKE PASS");
	return 0;
}

process.exit(await main());
